src/main/java/org/mule/extension/ftp/internal/source/FtpDirectoryListener.java

```diff
@@ -0,0 +1,461 @@
+
+  @Override
+  public void onStart(SourceCallback<InputStream, FtpFileAttributes> sourceCallback) throws MuleException {
+    matcher = predicateBuilder != null ? predicateBuilder.build() : new NullFilePayloadPredicate<>();
+
+    listenerExecutor = schedulerService.customScheduler(SchedulerConfig.config()
+        .withMaxConcurrentTasks(1)
+        .withName(
+                  format("%s.ftp.listener", location.getRootContainerName())));
+
+    directoryPath = resolveRootPath();
+    filesBeingProcessingObjectStore = objectStoreManager.getOrCreateObjectStore("ftp-listener:" + directory,
+                                                                                ObjectStoreSettings.builder()
+                                                                                    .persistent(false)
+                                                                                    .maxEntries(1000)
+                                                                                    .entryTtl(60000L)
+                                                                                    .expirationInterval(20000L)
+                                                                                    .build());
+
+    if (watermarkEnabled) {
+      watermarkObjectStore = objectStoreManager.getOrCreateObjectStore("ftp-listener-watermark:" + directory,
+                                                                       unmanagedPersistent());
+    }
+
+    long freq = poolingFrequencyTimeUnit.toMillis(pollingFrequency);
+    stopRequested.set(false);
+    listenerExecutor.scheduleAtFixedRate(() -> poll(sourceCallback), freq, freq, MILLISECONDS);
+  }
+
+  @OnSuccess
+  public void onSuccess(@ParameterGroup(name = POST_PROCESSING_GROUP_NAME) PostActionGroup postAction,
+                        SourceCallbackContext ctx) {
+    postAction(postAction, ctx);
+  }
+
+  @OnError
+  public void onError(@ParameterGroup(name = POST_PROCESSING_GROUP_NAME) PostActionGroup postAction,
+                      SourceCallbackContext ctx) {
+    if (postAction.isApplyPostActionWhenFailed()) {
+      postAction(postAction, ctx);
+    }
+  }
+
+  @OnTerminate
+  public void onTerminate(SourceCallbackContext ctx) {
+    try {
+      ctx.<Runnable>getVariable(FILE_RELEASER_VAR).ifPresent(Runnable::run);
+    } finally {
+      FtpFileSystem fileSystem = ctx.getConnection();
+      if (fileSystem != null) {
+        fileSystemProvider.disconnect(fileSystem);
+      }
+    }
+  }
+
+  private void poll(SourceCallback<InputStream, FtpFileAttributes> sourceCallback) {
+    if (isRequestedToStop()) {
+      return;
+    }
+
+    FtpFileSystem fileSystem;
+    try {
+      fileSystem = openConnection();
+    } catch (Exception e) {
+      LOGGER.error(format("Could not obtain connection while trying to poll directory '%s'. %s", directoryPath.toString(),
+                          e.getMessage()));
+      return;
```

Comment:
shouldn't a ConnectionException be propagated?

Reply:
no. If the task fails, it won't be rescheduled. Need to log it, swallow it and hpe for the best in the next poll